File: codeconvert_ai/input1.py
import numpy as np

halfgate = np.float(30.0)
bigT = np.float(120.0)
heps = np.float(0.01)

def expm(x):
    x = np.float(x)
    if abs(x) > 0.5:
        e = np.exp(x) - 1.0
    else:
        p = np.float(x)
        e = np.float(p)
        for i in range(2, 20):
            p = (p * x) / i
            e = e + p
            if abs(p) <= abs(e * np.finfo(float).eps):
                return np.float(e)

def expmm(x):
    x = np.float(x)
    if abs(x) > 0.5:
        e = np.exp(x) - 1.0 - x
    else:
        p = np.float(x * x * 0.5)
        e = np.float(p)
        for i in range(3, 26):
            p = (p * x) / i
            e = e + p
            if abs(p) <= abs(e * np.finfo(float).eps):
                return np.float(e)

def coshm(x):
    x = np.float(x)
    return np.float(2) * np.sinh(x * 0.5) ** 2

def sinhm(x):
    x = np.float(x)
    if abs(x) > 0.5:
        s = np.sinh(x) - x
    else:
        p = np.float(x ** 3 / 6)
        s = np.float(p)
        xx = np.float(x * x)
        for i in range(5, 20, 2):
            p = p * xx / (i * (i - 1.0))
            s = s + p
            if abs(p) <= abs(s * np.finfo(float).eps):
                return np.float(s)

def coshmm(x):
    x = np.float(x)
    xh = x * 0.5
    return sinhm(xh) * (2 * np.sinh(xh) + x)

def xcms(x):
    x = np.float(x)
    if abs(x) > 0.5:
        e = np.float(x * coshm(x) - sinhm(x))
    else:
        p = np.float(x ** 3 / 3)
        e = np.float(p)
        xx = x * x
        for i in range(2, 16):
            i2 = i * 2
            p = p * xx / (i2 * (i2 + 1))
            e = e + i * p
            if abs(p) <= abs(e * np.finfo(float).eps):
                return np.float(e)

def enbase_t(tspan, hspan):
    tspan = np.float(tspan)
    hspan = np.float(hspan)
    if tspan < 0.0:
        raise ValueError("In enbase_t; thspan must be positive")
    if hspan == 0.0:
        return 1.0
    return (hspan ** 2 / expmm(-tspan)) * 0.5


def tbnewton(nh, m, hgts, hs, hgtp, p, q):
    nit = int(12)

    nh = int(nh)
    m = int(m)
    hgts = np.float(hgts)
    hs = np.float(hs)
    hgtp = np.float(hgtp)
    p = np.float(p) 
    q = np.float(q) 
    gate = 2 * halfgate / bigT
    tr = hgtp * halfgate / bigT
    for i in range(nh):
        tee = hgts[i] * halfgate / bigT
        he = hs[i]
        it = 1
        while it <= 12:
            eval_tspline(m, tr, p, q, tee, hac, dhadt)
            if it == 1:
                dhdt[i] = dhadt / bigT
            if dhadt == 0:
                break
            dh = hac - he
            dt = -dh / dhadt
            if abs(dt) > gate:
                print("WARNING! In tbnewton; i,it,dt/gate =", i, it, dt / gate)
                break
            if abs(dh) < np.finfo(float).eps:
                dhdt[i] = dhadt / bigT
                break
            tee = tee + dt
            it = it + 1
        FF = it > 12
        if FF:
            print("In tbnewton; Newton iterations seem not to be converging at i=", i)
            print("tee,he,hac,heps,dhadt:", tee, he, hac, np.finfo(float).eps, dhadt)
        te[i] = tee

        return te, dhdt, FF

def ubnewton(nh, m, hgts, hs, hgtp, p, q):
    nit = int(12)

    nh = int(nh)
    m = int(m)
    hgts = np.float(hgts)
    hs = np.float(hs)
    hgtp = np.float(hgtp)
    p = np.float(p)
    q = np.float(q)

    gate = 2 * halfgate
    tr = hgtp * halfgate
    for i in range(nh):
        tee = hgts[i] * halfgate
        he = hs[i]
        it = 1
        while it <= 12:
            eval_uspline(m, tr, p, q, tee, hac, dhadt)  #NE CHECK
            if it == 1:
                dhdt[i] = dhadt
            if dhadt == 0:
                break
            dh = hac - he
            dt = -dh / dhadt
            if abs(dt) > gate:
                print("WARNING! In ubnewton; i,it,dt/gate =", i, it, dt / gate)
                break
            if abs(dh) < np.finfo(float).eps:
                dhdt[i] = dhadt
                break
            tee = tee + dt
            it = it + 1
        FF = it > 12
        if FF:
            print("In ubnewton; Newton iterations seem not to be converging at i=", i)
            print("tee,he,hac,heps,dhadt:", tee, he, hac, np.finfo(float).eps, dhadt)
        te[i] = tee

    return te, dhdt, FF

def fit_gtspline(n, xs, ys, on):
    m = 0
    xa = np.zeros(n)
    ya = np.zeros(n)
    qa = np.zeros(n)
    ja = np.zeros(n)
    for i in range(n):
        if on[i]:
            m = m + 1
            xa[m-1] = xs[i]
            ya[m-1] = ys[i]
    fit_tspline(m, xa[:m], ya[:m], qa[:m], ja[:m], en, FF)
    if FF:
        print("In fit_gtspline; failure flag raised at call to fit_tspline")
        return
    k = 0
    for i in range(n):
        if on[i]:
            k = k + 1
            q[i] = qa[k-1]
            j[i] = ja[k-1]
            yac[i] = ys[i]
        else:
            eval_tsplined(m, xa[:m], ya[:m], qa[:m], xs[i], yac[i], q[i])
            j[i] = 0

    return q, j, yac, en, FF

def fit_tspline(n, xs, p, q, j, en, FF):
    if n < 1:
        raise ValueError("In fit_tspline; size of data array must be positive")
    if n == 1:
        q[:] = 0
        j[:] = 0
        en = 0
        return
    for i in range(1, n):
        if xs[i-1] >= xs[i]:
            FF = True
            print("In fit_tspline; xs data must increase strictly monotonically")
            return
    qq = np.zeros((n, 2))
    difp = np.zeros(n-1)
    cpp = np.zeros(n-1)
    cqp = np.zeros(n-1)
    sumq = np.zeros(n-1)
    for i in range(n-1):
        ip = i + 1
        difp[i] = p[ip] - p[i]
        x = (xs[ip] - xs[i]) * 0.5
        ch = np.cosh(x)
        sh = np.sinh(x)
        xcmsx2 = xcms(x) * 2
        egg = x * sh / xcmsx2
        ehh = ch / (2 * sh)
        ccc = egg + ehh
        cpp[i] = ch / xcmsx2
        cqp[i] = -difp[i] * sh / xcmsx2
        qq[i, 0] = qq[i, 0] + ccc
        qq[ip, -1] = qq[ip, -1] + egg - ehh
        qq[ip, 0] = qq[ip, 0] + ccc
    qq[0, 0] = qq[0, 0] + 1
    qq[n-1, 0] = qq[n-1, 0] + 1
    q[:n-1] = -cqp
    q[n-1] = 0
    q[1:n] = q[1:n] - cqp
    ldltb(n, 1, qq)
    ltdlbv(n, 1, qq, q)
    sumq[:] = q[:n-1] + q[1:n]
    en = 0.5 * (np.dot(difp**2, cpp) + np.dot(sumq, cqp))
    sb = q[0]
    for i in range(n-1):
        ip = i + 1
        x = (xs[ip] - xs[i]) * 0.5
        xcmsx2 = xcms(x) * 2
        ch = np.cosh(x)
        sh = np.sinh(x)
        sap = (sh * sumq[i] - ch * difp[i]) / xcmsx2
        sa = sap + q[i]
        j[i] = sa - sb
        sb = sap + q[ip]
    j[n-1] = q[n-1] - sb

def int_tspline(n, xs, p, q, m):
    m[0] = 0
    e = 0
    for i in range(n-1):
        ip = i + 1
        x = (xs[ip] - xs[i]) * 0.5
        t2 = x * x * 0.5
        shx = np.sinh(x)
        chmx = coshm(x)
        shmx = sinhm(x)
        chmmx = coshmm(x)
        xcmsx = xcms(x)
        pa = (p[ip] + p[i]) * 0.5
        pd = (p[ip] - p[i]) * 0.5 / x
        qa = (q[ip] + q[i]) * 0.5
        qd = (q[ip] - q[i]) * 0.5 / shx
        c = qd
        a = pa - c * chmx
        d = (qa - pd) * x / xcmsx
        b = qa - d * chmx
        m[i] = e + a * x - b * t2 + c * shmx - d * chmmx
        e = e + 2 * (a * x + c * shmx)
    m[n-1] = e

def fit_guspline(n, xs, ys, on, q, j, yac, en, FF):
    m = 0
    xa = np.zeros(n)
    ya = np.zeros(n)
    qa = np.zeros(n)
    ja = np.zeros(n)
    for i in range(n):
        if on[i]:
            m = m + 1
            xa[m-1] = xs[i]
            ya[m-1] = ys[i]
    fit_uspline(m, xa[:m], ya[:m], qa[:m], ja[:m], en, FF)
    if FF:
        print("In fit_guspline; failure flag raised at call to fit_uspline")
        return
    k = 0
    for i in range(n):
        if on[i]:
            k = k + 1
            q[i] = qa[k-1]
            j[i] = ja[k-1]
            yac[i] = ys[i]
        else:
            eval_usplined(m, xa[:m], ya[:m], qa[:m], xs[i], yac[i], q[i])
            j[i] = 0

def fit_uspline(n, xs, p, q, j, en, FF):
    if n < 1:
        raise ValueError("In fit_uspline; size of data array must be positive")
    if n == 1:
        q[:] = 0
        j[:] = 0
        en = 0
        return
    for i in range(1, n):
        if xs[i-1] >= xs[i]:
            FF = True
            print("In fit_uspline; xs data must increase strictly monotonically")
            return
    qq = np.zeros((n, 2))
    difp = np.zeros(n-1)
    cpp = np.zeros(n-1)
    cqp = np.zeros(n-1)
    sumq = np.zeros(n-1)
    for i in range(n-1):
        ip = i + 1
        difp[i] = p[ip] - p[i]
        x = (xs[ip] - xs[i]) * 0.5
        ch = np.cosh(x)
        sh = np.sinh(x)
        xcmsx2 = xcms(x) * 2
        egg = x * sh / xcmsx2
        ehh = ch / (2 * sh)
        ccc = egg + ehh
        cpp[i] = ch / xcmsx2
        cqp[i] = -difp[i] * sh / xcmsx2
        qq[i, 0] = qq[i, 0] + ccc
        qq[ip, -1] = qq[ip, -1] + egg - ehh
        qq[ip, 0] = qq[ip, 0] + ccc
    qq[0, 0] = qq[0, 0] + 1
    qq[n-1, 0] = qq[n-1, 0] + 1
    q[:n-1] = -cqp
    q[n-1] = 0
    q[1:n] = q[1:n] - cqp
    ldltb(n, 1, qq)
    ltdlbv(n, 1, qq, q)
    sumq[:] = q[:n-1] + q[1:n]
    en = 0.5 * (np.dot(difp**2, cpp) + np.dot(sumq, cqp))
    sb = q[0]
    for i in range(n-1):
        ip = i + 1
        x = (xs[ip] - xs[i]) * 0.5
        xcmsx2 = xcms(x) * 2
        ch = np.cosh(x)
        sh = np.sinh(x)
        sap = (sh * sumq[i] - ch * difp[i]) / xcmsx2
        sa = sap + q[i]
        j[i] = sa - sb
        sb = sap + q[ip]
    j[n-1] = q[n-1] - sb

File: codeconvert_ai/input1.f90
function expm(x) result(e)!
real(dp),intent(in ):: x
real(dp)            :: e
real(dp):: p
integer :: i
if(abs(x)>o2)then
   e=exp(x)-u1
else
   p=x; e=p
   do i=2,19; p=p*x/i; e=e+p; if(abs(p)<=abs(e*eps))return; enddo
endif
end function expm


function expmm(x) result(e)!                                           [expmm]
real(dp),intent(in ):: x
real(dp)            :: e

real(dp):: p
integer :: i

if(abs(x)>o2)then
   e=exp(x)-u1-x
else
   p=x*x*o2; e=p
   do i=3,25; p=p*x/i; e=e+p; if(abs(p)<=abs(e*eps))return; enddo
endif
end function expmm

function coshm(x) result(c)
real(dp),intent(in ):: x
real(dp)            :: c

c=2*sinh(x*o2)**2
end function coshm


function sinhm(x) result(s)
real(dp),intent(in ):: x
real(dp)            :: s

real(dp):: p,xx
integer :: i

if(abs(x)>o2)then
   s=sinh(x)-x
else
   p=x**3/6;  s=p;  xx=x*x
   do i=5,19,2; p=p*xx/(i*(i-1)); s=s+p; if(abs(p)<=abs(s*eps))return; enddo
endif
end function sinhm

function coshmm(x) result(c)
real(dp),intent(in ):: x
real(dp)            :: c

real(dp)            :: xh

xh=x*o2
c=sinhm(xh)*(2*sinh(xh)+x)
end function coshmm


function xcms(x) result(e)

real(dp),intent(in ):: x
real(dp)            :: e

real(dp):: p,xx
integer :: i,i2

if(abs(x)>o2)then
   e=x*coshm(x)-sinhm(x)
else
   p=x**3/3;  e=p;  xx=x*x
   do i=2,15
      i2=i*2; p=p*xx/(i2*(i2+1)); e=e+i*p; if(abs(p)<=abs(e*eps))return
   enddo
endif
end function xcms


function enbase_t(tspan,hspan)result(r)
real(dp),intent(in ):: tspan,hspan
real(dp)            :: r

if(tspan<u0)stop 'In enbase_t; thspan must be positive'
if(hspan==u0)then; r=u1; return; endif
r=hspan**2/expmm(-tspan)*o2
end function enbase_t


subroutine tbnewton(nh,m,bigT,halfgate,hgts,hs,hgtp,p,q, te,dhdt,FF)
integer,               intent(in ):: nh,m
real(dp),              intent(in ):: bigT,halfgate
integer ,dimension(nh),intent(in ):: hgts
real(dp),dimension(nh),intent(in ):: hs
integer, dimension(m), intent(in ):: hgtp
real(dp),dimension(m) ,intent(in ):: p,q
real(dp),dimension(nh),intent(out):: dhdt, te
logical,               intent(out):: FF

integer,parameter    :: nit=12
real(dp),dimension(m):: tr
real(dp)             :: gate,tee,he,hac,dhadt,dh,dt
integer              :: i,it

gate=2*halfgate/bigT
tr=hgtp*halfgate/bigT
do i=1,nh
   tee=hgts(i)*halfgate/bigT
   he=hs(i)
!  Use Newton iterations to estimate the rescaled time, tee, at which the
!  height is he
   it = 1
   do while (it <= nit)
      call eval_tspline(m,tr,p,q, tee,hac,dhadt)
      if(it==1)dhdt(i)=dhadt/bigT
      if(dhadt==u0)exit
      dh=hac-he
      dt=-dh/dhadt
      if(abs(dt)>gate)then
         write(41,*) 'WARNING! In tbnewton; i,it,dt/gate = ',i,it,dt/gate
         exit
      endif
      if(abs(dh)<heps)then
         dhdt(i)=dhadt/bigT
         exit
      endif
      tee=tee+dt
      it = it + 1
   enddo
   FF=(it>nit)
   if(FF)then
      write(41,'("In tbnewton; Newton iterations seem not to be")')
      write(41,'("converging at i=",i3)'),i
      write(41,'("tee,he,hac,heps,dhadt:",5(1x,e11.4))'),tee,he,hac,heps,dhadt
   endif
   te(i) = tee
enddo
end subroutine tbnewton


subroutine ubnewton(nh,m,halfgate,hgts,hs,hgtp,p,q, te,dhdt,FF)
integer,               intent(in ):: nh,m
real(dp),              intent(in ):: halfgate
integer, dimension(nh),intent(in ):: hgts
real(dp),dimension(nh),intent(in ):: hs
integer, dimension(m), intent(in ):: hgtp
real(dp),dimension(m) ,intent(in ):: p,q
real(dp),dimension(nh),intent(out):: dhdt, te
logical,               intent(out):: FF

integer,parameter    :: nit=12
real(dp),dimension(m):: tr
real(dp)             :: gate,tee,he,hac,dhadt,dh,dt
integer              :: i,it

gate=2*halfgate
tr=hgtp*halfgate
do i=1,nh
   tee=hgts(i)*halfgate
   he=hs(i)
!  Use Newton iterations to estimate the rescaled time, tee, at which the
!  height is he
   it = 1
   do while (it <= nit)
      call eval_uspline(m,tr,p,q, tee,hac,dhadt)
      if(it==1)dhdt(i)=dhadt
      if(dhadt==u0)exit
      dh=hac-he
      dt=-dh/dhadt
      if(abs(dt)>gate)then
         write(41,*) 'WARNING! In ubnewton; i,it,dt/gate = ',i,it,dt/gate
         exit
      endif
      if(abs(dh)<heps)then
         dhdt(i)=dhadt
         exit
      endif
      tee=tee+dt
      it = it + 1
   enddo
   FF=(it>nit)
   if(FF)then
      write(41,'("In ubnewton; Newton iterations seem not to be")')
      write(41,'("converging at i=",i3)'),i
      write(41,'("tee,he,hac,heps,dhadt:",5(1x,e11.4))'),tee,he,hac,heps,dhadt
   endif
   te(i) = tee
enddo
end subroutine ubnewton


subroutine fit_gtspline(n,xs,ys,on,q,j,yac,en,FF)
integer,              intent(in ):: n
real(dp),dimension(n),intent(in ):: xs,ys
logical, dimension(n),intent(in ):: on
real(dp),dimension(n),intent(out):: q,j,yac
real(dp),             intent(out):: en
logical,              intent(out):: FF

real(dp),dimension(n):: xa,ya,qa,ja
integer              :: i,k,m

m=0
do i=1,n
   if(on(i))then; m=m+1; xa(m)=xs(i); ya(m)=ys(i); endif
enddo
call fit_tspline(m,xa(1:m),ya(1:m),qa(1:m),ja(1:m),en,FF)
if(FF)then
   write(41,*) 'In fit_gtspline; failure flag raised at call to fit_tspline'
   return
endif
k=0
do i=1,n
   if(on(i))then
      k=k+1
      q(i)=qa(k)
      j(i)=ja(k)
      yac(i)=ys(i)
   else
      call eval_tsplined(m,xa(1:m),ya(1:m),qa(1:m),xs(i), yac(i),q(i))
      j(i)=0
   endif
enddo
end subroutine fit_gtspline


subroutine fit_tspline(n,xs,p,q,j,en,FF)
use pmat2, only: ldltb, ltdlbv
integer,                intent(in ):: n
real(dp),dimension(  n),intent(in ):: xs,p
real(dp),dimension(  n),intent(out):: q,j
real(dp),               intent(out):: en
logical,                intent(out):: FF
!----------------------------------------------------------------------------
integer                   :: i,ip
real(dp)                  :: x,ch,sh,sa,sb,sap,ccc,xcmsx2,egg,ehh
real(dp),dimension(n-1)   :: difp,sumq,cpp,cqp
real(dp),dimension(n,-1:0):: qq ! <- Tridiagonal, stored as rows of nonupper
!=============================================================================
FF=F
if(n<1)stop 'In fit_tspline; size of data array must be positive'
if(n==1)then; q=0; j=0; en=0; return; endif
! apply a strict monotonicity check on the xs:
do i=2,n
   if(xs(i-1)>=xs(i)) then
      FF=T
      write(41,*) 'In fit_tspline; xs data must increase strictly monotonically'
      return
   end if
enddo
qq=0
do i=1,n-1
   ip=i+1
   difp(i)=p(ip)-p(i)
   x=(xs(ip)-xs(i))*o2 
   ch=cosh(x);  sh=sinh(x)
   xcmsx2=xcms(x)*2
   egg=x*sh/xcmsx2; ehh=ch/(2*sh)
   ccc=egg+ehh
   cpp(i)=ch/xcmsx2 
   cqp(i)=-difp(i)*sh/xcmsx2 
   qq(i,0)=qq(i,0)+ccc; qq(ip,-1)=qq(ip,-1)+egg-ehh; qq(ip,0)=qq(ip,0)+ccc
enddo
qq(1,0)=qq(1,0)+1
qq(n,0)=qq(n,0)+1

q(1:n-1)=-cqp; q(n)=0
q(2:n)=q(2:n)-cqp

call ldltb(n,1,qq) 
call ltdlbv(n,1,qq,q)
sumq=q(1:n-1)+q(2:n) 

en=o2*(dot_product(difp**2,cpp)+dot_product(sumq,cqp))
sb=q(1)
do i=1,n-1
   ip=i+1
   x=o2*(xs(ip)-xs(i))
   xcmsx2=xcms(x)*2
   ch=cosh(x);  sh=sinh(x)
   sap=(sh*sumq(i)-ch*difp(i))/xcmsx2
   sa=sap+q(i)
   j(i)=sa-sb
   sb  =sap+q(ip)
enddo
j(n)=q(n)-sb 
end subroutine fit_tspline


subroutine int_tspline(n,xs,p,q, m)
integer,              intent(in ):: n
real(dp),dimension(n),intent(in ):: xs,p,q
real(dp),dimension(n),intent(out):: m
!-----------------------------------------------------------------------------
real(dp):: a,b,c,d,e,t2,x,pa,pd,qa,qd,shx,chmx,shmx,chmmx,xcmsx
integer :: i,ip


e=u0
do i=1,n-1
   ip=i+1
   x=(xs(ip)-xs(i))*o2 !<- interval half-width
   t2=x*x*o2
   shx  =sinh  (x)
   chmx =coshm (x)
   shmx =sinhm (x)
   chmmx=coshmm(x)
   xcmsx=xcms  (x)
   pa=(p(ip)+p(i))*o2
   pd=(p(ip)-p(i))*o2/x
   qa=(q(ip)+q(i))*o2
   qd=(q(ip)-q(i))*o2/shx
   c=qd
   a=pa-c*chmx
   d=(qa-pd)*x/xcmsx
   b=qa-d*chmx
   m(i)=e+a*x -b*t2 +c*shmx -d*chmmx
   e=e+2*(a*x+c*shmx)
enddo
m(n)=e
end subroutine int_tspline


subroutine fit_guspline(n,xs,ys,on,q,j,yac,en,FF)
integer,              intent(in ):: n
real(dp),dimension(n),intent(in ):: xs,ys
logical, dimension(n),intent(in ):: on
real(dp),dimension(n),intent(out):: q,j,yac
real(dp),             intent(out):: en
logical,              intent(out):: FF

real(dp),dimension(n):: xa,ya,qa,ja
integer              :: i,k,m

m=0
do i=1,n
   if(on(i))then; m=m+1; xa(m)=xs(i); ya(m)=ys(i); endif
enddo
call fit_uspline(m,xa(1:m),ya(1:m),qa(1:m),ja(1:m),en,FF)
if(FF)then
   write(41,*) 'In fit_guspline; failure flag raised at call to fit_uspline'
   return
endif
k=0
do i=1,n
   if(on(i))then
      k=k+1
      q(i)=qa(k)
      j(i)=ja(k)
      yac(i)=ys(i)
   else
      call eval_usplined(m,xa(1:m),ya(1:m),qa(1:m),xs(i), yac(i),q(i))
      j(i)=0
   endif
enddo
end subroutine fit_guspline


subroutine fit_uspline(n,xs,p,q,j,en,FF)

